Data/ByteString/Lazy.hs

```diff
@@ -733,6 +800,21 @@ dropWhile f = dropWhile'
             n | n < S.length c -> Chunk (S.drop n c) cs
               | otherwise      -> dropWhile' cs
 
+-- | Similar to 'P.dropWhileEnd',
+-- drops the longest (possibly empty) suffix of elements
+-- satisfying the predicate and returns the remainder.
+--
+-- @'dropWhileEnd' p@ is equivalent to @'reverse' . 'dropWhile' p . 'reverse'@.
+--
+-- @since 0.11.2.0
+dropWhileEnd :: (Word8 -> Bool) -> ByteString -> ByteString
+dropWhileEnd f = dropWhileEnd'
+  where dropWhileEnd' Empty        = Empty
```

Comment:
I suggest the following strategy. Start with an empty accumulator. If you see `Chunk c bs`, check whether the last char of `c` satisfies `f`. If it does not, you can immediately return accumulated strings and `c` - without inspecting `bs`! If the last char of `c` satisfies `f`, append `c` to the accumulator and proceed with `bs`. Further, when you finally encounter `Empty`, remove from accumulator a tail satisfying `f` (which may span over several chunks) and return it.

Reply:
> 
> 
> Sorry, @3kyro, I got distracted by other things for longer than planned. Hope it was not too long.

Hi @Bodigrim , there's absolutely no problem. I'm trying to wrap my ahead around your suggestion, but I'm not up to the task. Considering a function
````haskell
go (Chunk c cs) acc
    | f (S.last c) = go cs (acc `append` fromStrict c)
    | otherwise = acc `append` fromStrict c
```` 
Assuming we start from the left, I fail to see how I can not inspect `cs` if `f (S.last c)`  is `False`. How do we know that there isn't a `Word8` in `cs` that does not satisfy `f`? 

You also mention
> when you finally encounter Empty, remove from accumulator a tail satisfying f (which may span over several chunks) and return it.

doesn't this need an implenatation of `dropWhileEnd` that we are in the process of writing?

I'm certainly missing something obvious here so I would really appreciate your help.

Reply:
Sorry, I was not talking clear. I meant something like this:
```haskell
go :: [S.ByteString] -> ByteString -> ByteString
go acc (Chunk c cs) 
  | f (S.last c) = go (c : acc) cs 
  | otherwise    = foldl (flip Chunk) (c : acc) (go [] cs)
```
In the second case we can start constructing lazy `ByteString` before forcing `go [] cs`. This is important, because if a consumer cares only about the beginning of the string (e. g., `take n . dropWhileEnd f`), `go [] cs` could remain unevaluated.

You are absolutely right, for `go acc Empty` we basically employ an eager algorithm you have written for `dropWhile` above (note that chunks in `acc` are in reversed order, however).

Reply:
Makes absolute sense, thanks a lot.

I've added your suggestions plus a lazy test for `dropWhileEnd`

Reply:
`breakEnd` and `spanEnd` updated